addons/eu/en16931/bill.go

```diff
@@ -149,37 +162,80 @@ func validateBillLineDiscount(value any) error {
 	if !ok || discount == nil {
 		return nil
 	}
-	if discount.Reason == "" && !discount.Ext.Has(untdid.ExtKeyAllowance) {
-		return validation.NewError("BR-41", "either a reason or an allowance type extension is required")
-	}
-	return nil
+
+	return validation.ValidateStruct(discount,
+		// BR-41
+		validation.Field(&discount.Reason,
+			validation.When(
+				!discount.Ext.Has(untdid.ExtKeyAllowance),
+				validation.Required.Error("either a reason or an allowance type extension is required (BR-41)"),
+			),
+			validation.Skip,
+		),
+		validation.Field(&discount.Ext,
+			validation.When(
+				discount.Reason == "",
+				validation.Required.Error("either a reason or an allowance type extension is required (BR-41)"),
+			),
+			validation.Skip,
+		),
+	)
 }
 
 func validateBillCharge(charge *bill.Charge) error {
 	// BR-36
-	if charge.Reason == "" && (charge.Ext == nil || charge.Ext[untdid.ExtKeyCharge] == "") {
-		return validation.NewError("BR-36", "either a reason or a charge type extension is required")
-	}
-	return nil
+	return validation.ValidateStruct(charge,
+		validation.Field(&charge.Reason,
+			validation.When(
+				!charge.Ext.Has(untdid.ExtKeyCharge),
+				validation.Required.Error("either a reason or a charge type extension is required (BR-36)"),
+			),
+			validation.Skip,
+		),
+		validation.Field(&charge.Ext,
+			validation.When(
+				charge.Reason == "",
+				validation.Required.Error("either a reason or a charge type extension is required (BR-36)"),
+			),
+			validation.Skip,
+		),
+	)
 }
 
 func validateBillDiscount(discount *bill.Discount) error {
 	// BR-33
-	if discount.Reason == "" && (discount.Ext == nil || discount.Ext[untdid.ExtKeyAllowance] == "") {
-		return validation.NewError("BR-33", "either a reason or an allowance type extension is required")
-	}
-	return nil
+	return validation.ValidateStruct(discount,
+		validation.Field(&discount.Reason,
+			validation.When(
+				!discount.Ext.Has(untdid.ExtKeyAllowance),
+				validation.Required.Error("either a reason or an allowance type extension is required (BR-33)"),
+			),
+			validation.Skip,
+		),
+		validation.Field(&discount.Ext,
+			validation.When(
+				discount.Reason == "",
+				validation.Required.Error("either a reason or an allowance type extension is required (BR-33)"),
+			),
+			validation.Skip,
+		),
+	)
 }
 
-func validateBillPayment(payment *bill.PaymentDetails) error {
-	if payment == nil {
+func validateBillPayment(value any) error {
+	payment, ok := value.(*bill.PaymentDetails)
+	if !ok || payment == nil {
 		return nil
 	}
-
 	return validation.ValidateStruct(payment,
 		validation.Field(&payment.Terms,
-			validation.Required,
-			validation.Skip,
+			validation.Required.Error("payment terms are required when amount is due (BR-CO-25)"),
 		),
 	)
 }
+
+func isDue(inv *bill.Invoice) bool {
```

Comment:
I suggest using a more complete name, despite being a private method, to avoid potential for conflicts in the future:

```suggestion
func billInvoicePaymentIsDue(inv *bill.Invoice) bool {
```